zilliqa/src/bin/validator_oracle.rs

```diff
@@ -0,0 +1,238 @@
+use std::{path::PathBuf, str::FromStr};
+
+use alloy_provider::{Provider, ProviderBuilder, WsConnect};
+use alloy_rpc_types::{BlockNumberOrTag, Filter};
+use anyhow::Result;
+use clap::Parser;
+use ethers::{
+    contract::abigen,
+    providers::Middleware,
+    signers::{LocalWallet, Signer},
+    types::{TransactionRequest, H160},
+};
+use futures_util::stream::StreamExt;
+use tokio::sync::watch;
+use tracing::{debug, error, info};
+use tracing_subscriber::EnvFilter;
+use zilliqa::{
+    contracts,
+    crypto::{NodePublicKey, SecretKey},
+    state::contract_addr,
+    uccb::{
+        cfg::{ChainConfig, Config, ZQ2Config},
+        client::ChainClient,
+    },
+};
+
+// The json files are generated by invoking 'forge build' in zilliqa/src/contracts/uccb
+abigen!(
+    ValidatorManager,
+    "zilliqa/src/contracts/uccb/out/ValidatorManager.sol/ValidatorManager.json"
+);
+
+#[derive(Parser, Debug)]
+struct Args {
+    #[arg(value_parser = SecretKey::from_hex)]
+    secret_key: SecretKey,
+    #[clap(long, short, default_value = "config.toml")]
+    config_file: PathBuf,
+}
+
+impl zilliqa::uccb::Args for Args {
+    fn secret_key(&self) -> &SecretKey {
+        &self.secret_key
+    }
+
+    fn config_file(&self) -> &PathBuf {
+        &self.config_file
+    }
+}
+
+struct ValidatorOracle {
```

Comment:
As a bit of a meta point, I'm finding it hard to follow the logic of the validator oracle since it is split across multiple files (which are quite far apart in the directory tree). Since the validator oracle is quite simple, I'd be in favour of just keeping it in one file for now?

Reply:
These files will be shared by the uccb logic so I won't put everything here just to move them out later. The whole idea is that it's in its separate module (as much as possible). Plus, I suppose I prefer it the other way: I dislike very long files with tons of (sometimes unrelated) logic in it.